### DISCUSSION TITLE
Is it possible to use the tracing crate within handlers?

### DISCUSSION THREAD

**Discussion by harrytucker** [rwf2/Rocket#1673] [Questions] [answered]
First off thanks for all the work that everyone put into 0.5, I'm playing around with it and it looks fantastic 😄 

In 0.5 I understand that Rocket's logging infrastructure has been migrated to the `tracing` crate. After skimming through the API docs for logging items I can't find any examples of logging. This minimal example results in no logging output from the `info!` call within the handler:
```rust
#[macro_use] extern crate rocket;
#[macro_use] extern crate tracing;

#[get("/<name>/<age>")]
fn hello(name: &str, age: u8) -> String {
    info!(
        ?name,
        ?age,
        "Processing request",
    );
    format!("Hello, {} year old named {}!", age, name)
}

#[launch]
fn rocket() -> _ {
    rocket::build().mount("/hello", routes![hello])
}
```

I have an application built on actix that I'd like to port over to the new release which uses `tracing` throughout, am I missing something obvious in the docs? Or is it not currently possible to hook into Rocket's logging facilities?

**Comment by SergioBenitez** [comment, marked as answer]
> In 0.5 I understand that Rocket's logging infrastructure has been migrated to the `tracing` crate.

This unfortunately isn't the case yet. #21 and #1579 are the latest on moving over to `tracing`. For now, Rocket still uses `log`. We don't document how to tap in in hopes of being able to migrate to `tracing` in a non-breaking release. But in a pinch, using `log` macros will work and should continue to work even when we do migrate to `tracing`.

**Reply by harrytucker** [reply]
Ah cool, thanks for the heads-up! 👍

**Reply by SergioBenitez** [reply]
See also this part of @jebrosen's response:

> In the meantime, tracing can [emit `log` events](https://docs.rs/tracing/0.1.26/tracing/#emitting-log-records) in addition to `tracing` events; these events would be seen by Rocket's default logger. Another possible option is to register your own `tracing` subscriber that [captures `log` events](https://docs.rs/tracing/0.1.26/tracing/#consuming-log-records) (including those emitted by `rocket`) before rocket's logger can be initialized. Neither of these options are thoroughly supported and there are a few wrinkles; nonetheless, they might be a suitable choice in some situations.

**Comment by jebrosen** [comment]
> In 0.5 I understand that Rocket's logging infrastructure has been migrated to the tracing crate.

Unfortunately I did not complete this feature in time for 0.5 after all; issue #21 / PR #1579 are still open. At present only `log` events are captured by Rocket's own logging mechanisms, although this is not documented as part of the API.

In the meantime, tracing can [emit `log` events](https://docs.rs/tracing/0.1.26/tracing/#emitting-log-records) in addition to `tracing` events; these events would be seen by Rocket's default logger. Another possible option is to register your own `tracing` subscriber that [captures `log` events](https://docs.rs/tracing/0.1.26/tracing/#consuming-log-records) (including those emitted by `rocket`) before rocket's logger can be initialized. Neither of these options are thoroughly supported and there are a few wrinkles; nonetheless, they might be a suitable choice in some situations.

EDIT: Ninja'd!
